presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java

```diff
@@ -693,4 +697,37 @@ public static long widthBucket(@SqlType(StandardTypes.DOUBLE) double operand, @S
 
         return lower;
     }
+
+    @ScalarFunction
+    @SqlType(StandardTypes.DOUBLE)
+    public static double cosineSimilarity(
+            @SqlType(MAP_STRING_DOUBLE) Block leftMap,
+            @SqlType(MAP_STRING_DOUBLE) Block rightMap)
+    {
+        double dotProduct = dotProduct(leftMap, rightMap);
+        double normLeftMap = Math.sqrt(dotProduct(leftMap, leftMap));
+        double normRightMap = Math.sqrt(dotProduct(rightMap, rightMap));
+
+        return dotProduct / (normLeftMap * normRightMap);
+    }
+
+    private static double dotProduct(Block leftMap, Block rightMap)
+    {
+        TypedSet keysN = new TypedSet(VarcharType.VARCHAR, rightMap.getPositionCount());
+
+        for (int i = 0; i < rightMap.getPositionCount(); i += 2) {
+            keysN.add(rightMap, i);
+        }
+
+        double result = 0.0;
+
+        for (int i = 0; i < leftMap.getPositionCount(); i += 2) {
+            if (keysN.contains(leftMap, i)) {
```

Comment:
This ends up doing two lookups. It should be possible to do this once with some changes to the `TypedSet` API (similar to how a standard `Map` returns `null` if not found)